Frame.cc

```diff
@@ -1153,3 +1204,95 @@ bool Frame::GetSpectralData(std::vector<float>& data, casacore::SubImage<float>&
+    std::vector<std::vector<double>> results(stats_size);
+    for (int i = 0; i < stats_size; ++i) {
+        results[i].resize(profile_size, std::numeric_limits<double>::quiet_NaN());
+    }
+    // get region state for this process
+    RegionState region_state = region->GetRegionState();
+    // get statistical profile data
+    int start = 0;
+    int count, end;
+    float progress = 0;
+    casacore::SubImage<float> sub_image;
+    while (start < profile_size) {
+        // start the timer
+        auto tStart = std::chrono::high_resolution_clock::now();
+        // check if frontend queries changed, if so, terminate this loop process
+        if (!_connected) {
+            std::cerr << "[Region " << region_id << "] closing image, exit zprofile (statistics) before complete" << std::endl;
+            return false;
+        }
+        if (_region_states.count(region_id) && _region_states[region_id] != region_state) {
+            std::cerr << "[Region " << region_id << "] region state changed, exit zprofile (statistics) before complete" << std::endl;
+            return false;
+        }
+        if (_region_configs.count(region_id) && !_region_configs[region_id].IsAmong(profile_index, requested_stats)) {
+            std::cerr << "[Region " << region_id << "] region requirement changed, exit zprofile (statistics) before complete" << std::endl;
+            return false;
+        }
+        end = (start + delta_channels > profile_size ? profile_size - 1 : start + delta_channels - 1);
+        count = end - start + 1;
+        GetRegionSubImage(region_id, sub_image, profile_stokes, ChannelRange(start, end));
+        std::vector<std::vector<double>> buffer;
+        if (region->GetSpectralProfileData(buffer, profile_index, sub_image)) {
+            for (int j = 0; j < stats_size; ++j) {
+                memcpy(&results[j][start], &buffer[j][0], count * sizeof(double));
+            }
+        } else {
+            std::cerr << "Can not get zprofile (statistics), region id: " << region_id
+                      << ", channel range: ["<< start << "," << end << "]" << std::endl;
+            return false;
+        }
+        start += count;
+        progress = (float)start / profile_size;
+        // get the time elapse for this step
+        auto tEnd = std::chrono::high_resolution_clock::now();
+        auto dt = std::chrono::duration<double, std::milli>(tEnd - tStart).count();
+        // adjust the increment of channels according to the time elapse
+        delta_channels *= dt_target / dt;
+        if (delta_channels < 1) {
+            delta_channels = 1;
+        }
+        if (delta_channels > profile_size) {
+            delta_channels = profile_size;
+        }
+        // send partial result by the callback function
+        cb(results, progress);
+    }
+    stats_values = std::move(results);
+    return true;
+}
+
+bool Frame::CanUseSiwzzledData(const casacore::ArrayLattice<casacore::Bool>* mask) {
```

Comment:
Typo in the name. I would also rename this to something like `UseImageRegionStats` -- this function specifically checks if the swizzled dataset can and should be used for this specific application, not if it exists in general.